### ISSUE TITLE
AttributeError: 'WorkerStore' object has no attribute 'request_dispatcher' on CHANNEL_HTTP_SOAP_CREATE_EDIT

### ISSUE THREAD

**Issue by wbar** [zatosource/zato#395]
When creating SOAP CHANNEL:

```
2015-03-16 16:23:47,766 - ERROR - 24956:Dummy-3 - zato.server.base:22 - Could not handle broker msg:[Bunch(action=u'101012', cluster_id=1L, connection=u'channel', data_format=u'', has_rbac=u'', host=u'', id=540L, impl_name=u'mock_authn_server.MockAuthServer', is_active=True, is_internal=False, merge_url_params_req=True, method=u'GET', msg_type=u'0002', name=u'urn:dtag:falcon-esb:stub:authn:info', old_name=u'urn:dtag:falcon-esb:stub:authn:info', old_soap_action=u'', old_url_path=u'/oauth/token/info', params_pri=u'channel-params-over-msg', ping_method=u'', pool_size=u'', sec_tls_ca_cert_id=u'', sec_type=None, security_id=None, security_name=None, serialization_type=u'', service=u'urn:dtag:falcon-esb:stub:authn_info', service_id=1073L, service_name=u'urn:dtag:falcon-esb:stub:authn_info', soap_action=u'', soap_version=u'', timeout=u'', transport=u'plain_http', url_params_pri=u'qs-over-path', url_path=u'/oauth/token/info')], e:[Traceback (most recent call last):
  File "/opt/zato/2.0.2/zato-server/src/zato/server/base/__init__.py", line 47, in on_broker_msg
    getattr(self, handler)(msg)
  File "/opt/zato/2.0.2/zato-server/src/zato/server/base/worker.py", line 1069, in on_broker_msg_CHANNEL_HTTP_SOAP_CREATE_EDIT
    self.request_dispatcher.url_data.on_broker_msg_CHANNEL_HTTP_SOAP_CREATE_EDIT(msg, *args)
AttributeError: 'WorkerStore' object has no attribute 'request_dispatcher'
]
````

**Comment by dsuch** [comment]
I cannot reproduce it. What is the set of parameters that you are providing the web-admin with? Are you using web-admin? What browser is it?

**Comment by wbar** [comment]
I reproduced an error. This happend when server is still booting up and request has been send. Maybe server should report it self into loadbalancer when finishes booting ?

**Comment by wbar** [comment]
I've noticed that this error not happening when I see something like this in server log:
```
2015-03-16 19:45:13,273 - INFO - 1481:Dummy-163 - zato.server.base.parallel:22 - Initializing connectors
2015-03-16 19:45:13,283 - INFO - 1481:Dummy-163 - zato.server.base.parallel:22 - No AMQP channels to start
2015-03-16 19:45:13,288 - INFO - 1481:Dummy-163 - zato.server.base.parallel:22 - No AMQP outgoing connections to start
2015-03-16 19:45:13,294 - INFO - 1481:Dummy-163 - zato.server.base.parallel:22 - No JMS WebSphere MQ channels to start
2015-03-16 19:45:13,299 - INFO - 1481:Dummy-163 - zato.server.base.parallel:22 - No JMS WebSphere MQ outgoing connections to start
2015-03-16 19:45:13,305 - INFO - 1481:Dummy-163 - zato.server.base.parallel:22 - No Zero MQ channels to start
```
How wait for this ?
I don't want to use any bash based solution

**Comment by dsuch** [comment]
Hi @wbar

* When a server is guaranteed to be fully initialized it can invoke startup services of your choice.
* That service can perform arbitrary operations, like notifying you through some means that it has just become ready

Perhaps this is the right approach in your situation? Instead of waiting, be notified when a server can be used?

Startup services are configured in server.conf:

https://zato.io/docs/admin/guide/install-config/config-server.html#startup-services-first-worker

**Comment by wbar** [comment]
Thx @dsuch for reply, but I've noticed something, possibly, terryfing ...

In log I can see something like this:
```
2015-03-16 20:25:54,854 - INFO - 14523:Dummy-8 - zato.helpers.input-logger:784 - {u'impl_name': u'zato.server.service.internal.helpers.InputLogger', u'name': u'zato.helpers.input-logger', u'cid': u'K06KQVEW24KZ6GSDRV7A3FFDV1WJ', u'invocation_time': datetime.datetime(2015, 3, 16, 20, 25, 54, 853492), u'job_type': None, u'data_format': None, u'slow_threshold': 99999, u'request.payload': u'Sample payload for a startup service (first worker)', u'wsgi_environ': {u'zato.request_ctx.async_msg': Bunch(action=u'101802', channel=u'startup-service', cid=u'K06KQVEW24KZ6GSDRV7A3FFDV1WJ', msg_type=u'0001', payload=u'Sample payload for a startup service (first worker)', service=u'zato.helpers.input-logger'), u'zato.request_ctx.fanout_cid': None, u'zato.request_ctx.in_reply_to': None, u'zato.request_ctx.parallel_exec_cid': None}, u'environ': {}, u'usage': 99, u'channel': u'startup-service'}
```
And I think that is from first worker when second one is booting. After second one boots up with log entry:
```
2015-03-16 20:26:13,742 - INFO - 14523:Dummy-9 - zato.helpers.input-logger:784 - {u'impl_name': u'zato.server.service.internal.helpers.InputLogger', u'name': u'zato.helpers.input-logger', u'cid': u'K05NZ3SXF1HVSK0TVWW98R9T0NP4', u'invocation_time': datetime.datetime(2015, 3, 16, 20, 26, 13, 740860), u'job_type': None, u'data_format': None, u'slow_threshold': 99999, u'request.payload': u'Sample payload for a startup service (any worker)', u'wsgi_environ': {u'zato.request_ctx.async_msg': Bunch(action=u'101802', channel=u'startup-service', cid=u'K05NZ3SXF1HVSK0TVWW98R9T0NP4', msg_type=u'0001', payload=u'Sample payload for a startup service (any worker)', service=u'zato.helpers.input-logger'), u'zato.request_ctx.fanout_cid': None, u'zato.request_ctx.in_reply_to': None, u'zato.request_ctx.parallel_exec_cid': None}, u'environ': {}, u'usage': 100, u'channel': u'startup-service'}
```
everything works perfectly fine. I can invoke services and everything is perfect.

But...

When first worker bootup and the second one is still booting (I think) sometimes `master` process send job into that second worker and I can see log entry like this:

```
2015-03-16 20:14:38,073 - ERROR - 12233:Dummy-9 - zato.server.base:22 - Could not handle broker msg:[Bunch(action=u'102200', cid=u'K051CDJMHYCMS66N2EXMAPQK7MJS', data_format=u'json', msg_type=u'0002', package_id=4640L, payload={u'package_id': 4640L}, service=u'zato.hot-deploy.create')], e:[Traceback (most recent call last):
  File "/opt/zato/2.0.2/zato-server/src/zato/server/base/__init__.py", line 47, in on_broker_msg
    getattr(self, handler)(msg)
  File "/opt/zato/2.0.2/zato-server/src/zato/server/base/worker.py", line 1187, in on_broker_msg_HOT_DEPLOY_CREATE
    return self._on_message_invoke_service(msg, 'hot-deploy', 'HOT_DEPLOY_CREATE', args)
  File "/opt/zato/2.0.2/zato-server/src/zato/server/base/worker.py", line 1003, in _on_message_invoke_service
    environ=msg.get('environ'))
  File "/opt/zato/2.0.2/zato-server/src/zato/server/service/__init__.py", line 389, in update_handle
    environ=kwargs.get('environ'))
  File "/opt/zato/2.0.2/zato-server/src/zato/server/service/__init__.py", line 830, in update
    service._init()
  File "/opt/zato/2.0.2/zato-server/src/zato/server/service/__init__.py", line 273, in _init
    out_sql = self.worker_store.sql_pool_store
AttributeError: 'WorkerStore' object has no attribute 'sql_pool_store'
```

**Comment by dsuch** [comment]
> line 1187, in on_broker_msg_HOT_DEPLOY_CREATE

Ok, but who is hot-deploying code here? Are these your deployment scripts?

The whole flow is then:

* First worker starts
* You deploy
* Another worker, not yet started, picks up the hot-deploy message and fails

?

**Comment by wbar** [comment]
We are using deployment service

**Comment by dsuch** [comment]
Ok, but I still don't know how to reproduce it.

Can you confirm that the steps I listed are what is needed to reliably reproduce it?

**Comment by wbar** [comment]
the whole flow looks like ok, but second worker need to be in booting stage. I think that it subscribes to mster before setting a connection pool with ODB and/or KVDB

**Comment by dsuch** [comment]
@wbar - the system it is all happening on, how fast it is? Is it a slow VM by any chance?

**Comment by dsuch** [comment]
Also @wbar - please send the full contents of server.log from the moment the server starts to boot to the moment the error is logged. Thanks.

**Comment by dsuch** [comment]
Hi @wbar - I have reproduced and fixed it.

The reason was that a worker connected to Redis before it read all of its configuration from SQL ODB.

When you deploy new services or update channels, the messages is delivered to a worker through Redis.

So the situation was that even though a worker wasn't fully initialized it would receive a message from Redis which ended up in an exception being logged.

This is fixed in git and will be released in 2.0.3.

However, in your situation, you still need to be careful and you cannot send configuration messages too soon, when the worker is still booting because now they won't be received by the worker at all - it's still booting after all and its Redis client didn't connect yet.

You have two options:

* Wait arbitrary amount of time before creating channels and deployment (less than ideal)
* Configure [startup_services_first_worker](https://zato.io/docs/admin/guide/install-config/config-server.html#startup-services-first-worker) and [startup_services_any_worker](https://zato.io/docs/admin/guide/install-config/config-server.html#startup-services-any-worker) to point to your own services that will signal to you that a server and all its workers are fully initialized

**Comment by dsuch** [comment]
Done and released in 2.0.3.
